src/client/dfs/dfs.c

```diff
@@ -2350,6 +2350,100 @@ dfs_chmod(dfs_t *dfs, dfs_obj_t *parent, const char *name, mode_t mode)
 	return rc;
 }
 
+int
+dfs_osetstat(dfs_t *dfs, dfs_obj_t *obj, struct stat *stbuf, int flags)
+{
+	daos_handle_t		th = DAOS_TX_NONE;
+	uid_t			euid;
+	daos_key_t		dkey;
+	daos_handle_t           oh;
+	int			rc;
+	d_sg_list_t		sgls[3];
+	d_iov_t			sg_iovs[3];
+	daos_iod_t		iods[3];
+	int i = 0;
+	int akeys_nr;
+
+	if (dfs == NULL || !dfs->mounted)
+		return -EINVAL;
+	if (obj == NULL)
+		return -EINVAL;
+	if (dfs->amode != O_RDWR)
+		return -EPERM;
+
+	euid = geteuid();
+	/** only root or owner can change mode */
+	if (euid != 0 && dfs->uid != euid)
+		return -EPERM;
+
+	/** Open parent object and fetch entry of obj from it */
+	rc = daos_obj_open(dfs->coh, obj->parent_oid, DAOS_OO_RO, &oh, NULL);
+	if (rc)
+		return -daos_der2errno(rc);
+
+	/** set dkey as the entry name */
+	d_iov_set(&dkey, (void *)obj->name, strlen(obj->name));
+
+	if (flags & DFS_SET_ATTR_MODE) {
+		/** set akey as the mode attr name */
+		d_iov_set(&sg_iovs[i], &stbuf->st_mode, sizeof(mode_t));
+		d_iov_set(&iods[i].iod_name, MODE_NAME, strlen(MODE_NAME));
+		iods[i].iod_size = sizeof(mode_t);
+		i++;
+		flags &= ~DFS_SET_ATTR_MODE;
+	}
+	if (flags & DFS_SET_ATTR_ATIME) {
+		d_iov_set(&sg_iovs[i], &stbuf->st_atim, sizeof(time_t));
+		d_iov_set(&iods[i].iod_name, ATIME_NAME, strlen(ATIME_NAME));
+		iods[i].iod_size = sizeof(time_t);
+
+		flags &= ~DFS_SET_ATTR_ATIME;
+		i++;
+	}
+	if (flags & DFS_SET_ATTR_MTIME) {
+		d_iov_set(&sg_iovs[i], &stbuf->st_mtim, sizeof(time_t));
+		d_iov_set(&iods[i].iod_name, MTIME_NAME, strlen(MTIME_NAME));
+		iods[i].iod_size = sizeof(time_t);
+
+		flags &= ~DFS_SET_ATTR_MTIME;
+		i++;
+	}
+
+	if (flags) {
+		D_GOTO(out, rc = -EINVAL);
+	}
+
+	akeys_nr = i;
+
+	for (i = 0; i < akeys_nr; i++) {
+		sgls[i].sg_nr		= 1;
+		sgls[i].sg_nr_out	= 0;
+		sgls[i].sg_iovs		= &sg_iovs[i];
+
+		daos_csum_set(&iods[i].iod_kcsum, NULL, 0);
+		iods[i].iod_nr		= 1;
+		iods[i].iod_recxs	= NULL;
+		iods[i].iod_eprs	= NULL;
+		iods[i].iod_csums	= NULL;
+		iods[i].iod_type	= DAOS_IOD_SINGLE;
+	}
+
+	rc = daos_obj_update(oh, th, &dkey, akeys_nr, iods, sgls, NULL);
+	if (rc) {
+		D_ERROR("Failed to update attr (rc = %d)\n", rc);
+		D_GOTO(out, rc = -daos_der2errno(rc));
+	}
+
+	rc = entry_stat(dfs, th, oh, obj->name, stbuf);
```

Comment:
Ideally this would all be read in the same transaction, otherwise there's a potential race condition with rename that we'd need to check for and decide how to handle.  I did initially implement it as needing a separate ostat() call and can move it back to that.

Reply:
right, if we want to do the revalidate before the access, we would need to read the dkey (probably just the oid, not the whole entry), and verify that with the oid from the object. But we can read the whole dkey and in that case we can return the stat, however for the weaker consistency mode, we don't want to revalidate since we won't use transactions there (in that case you can validate the oid, rename happens, then you do setattr, and if you don't use tx, it won't be detected).
This should apply to other APIs like ostat itself, all the xattr, etc.

So i guess what im saying, we don't want to impose through the API this revalidate step. But if you think in most use cases, a stat is required after a setattr, we can make it in/out with an optional flag to stat or not. in that case i would still call the function dfs_osetattr.

Reply:
Look at the rest of the PR, there's only one caller to d_osetstat() and it needs the attr afterwards.

This interface itself doesn't enforce use of revalidate, but gives a performance improvement if it is being done as more data would be read but it would be the same number of round-trips.

Calling ostat() afterwards where the ostat itself did a revalidate would also work, and would allow us to detect the race condition, but would not allow us to protect against it.

Reply:
OK got it, let's keep the API as is, and we could investigate the revalidate when transactions are done.
the name of the function is still weird in my opinion, and setattr() would be better ;-)

Reply:
Will change, I don't like the current name either.